### PR TITLE
Geomap: Add dynamic initial view options

### LINKED ISSUE #40496: [ feature request ] [ GeoMap ] dynamically zoom to location based on input
automictic zoom to location based on data content will be great value add .

### PR DESCRIPTION
**Pull request by drew08t**
What this PR does / why we need it:
For dynamic data that changes, setting up map view to "fit data layers" does not update upon data change. As an extension to this support, minimum control options should also be included:

- [x] - Max zoom
- [x] - Padding (%)
- [x] - Last data point only (boolean)
- [x] - Handle single layer bug

After:
Aug-31-2022 13-06-22

Closes #40496

### CODE REVIEW

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (drew08t) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:597]: remove this.

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (drew08t) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:607]: remove this.

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (drew08t) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:271]: Should add conditional for data fit, so this view init only gets called for that view setting.

File: public/app/plugins/panel/geomap/editor/MapViewEditor.tsx
Comment (nmarrs) [inline public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:116]: This would probably make sense to split out into an individual component for improved maintainability / testing

File: public/app/plugins/panel/geomap/editor/MapViewEditor.tsx
Comment (nmarrs) [inline public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:135]: For values it might make sense to create an enum type for better maintainability

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (nmarrs) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:615]: any insight on this calculation? (i.e. why padding / 100 + 1)

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (nmarrs) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:607]: We could also just pass `config.lastOnly` here and set the default param to `false` in `getLayerExtent` definition
```
export function getLayersExtent(layers: Collection<BaseLayer>, lastOnly: boolean = false): Extent {
```

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (drew08t) [reply public/app/plugins/panel/geomap/GeomapPanel.tsx:617]: Padding is a percentage > 0, so this is a multiplier to adjust the resolution accordingly. 50% for example would multiply resolution by 1.5 = 50 / 100 + 1.

File: public/app/plugins/panel/geomap/editor/MapViewEditor.tsx
Comment (drew08t) [reply public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:120]: Will look into that after resolving conflicts on GeomapPanel.

File: public/app/plugins/panel/geomap/editor/MapViewEditor.tsx
Comment (drew08t) [reply public/app/plugins/panel/geomap/editor/MapViewEditor.tsx]: Agreed.

**Comment by drew08t** [comment]
New options controls for data fit:
Sep-14-2022 09-46-28

File: public/app/plugins/panel/geomap/editor/MapViewEditor.tsx
Comment (nmarrs) [inline public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:133]: It might be nice to break up this component up into separately defined fragments for better readability / maintainability (not just additions here but the entire component)

**Review by nmarrs — COMMENTED** [review]
Overall this is looking great! I think we are getting close

A few things I noticed:
- We do not update the view when when switching layer types - maybe this isn't necessary though given we only really care about the underlying data updating -> though given current behavior with geojson / heatmap layers this may be a nice addition depending on where we land there
- Streaming data doesn't seem to work -> though that may be due to other issues (table view doesn't seem to work when streaming so there is no data to fit to?)
- This feature doesn't seem to work as expected for some layer types (heatmap, geojson) (see screencast below)


https://user-images.githubusercontent.com/22381771/191168102-6ce55f8c-0f50-486d-a145-b795604a43e6.mov

**Comment by nmarrs** [comment]
Looked this over with Ryan and came to following conclusions:
For now heat map is a little hard to address given the nature of how heat map works, for now we can push this off

As for re-initializing view on layer type change, we can also hold off on that as well. Also, we probably will want to do some exploration for how initial view actually works (whether or not local storage is utilized) etc.

Streaming was a local issue -> It appears the `live-service-web-worker` feature flag seems to break streaming right now (current solution is to just disable / remove this feature flag and streaming works as expected)

**Review by nmarrs — APPROVED** [review]
This is looking good to me as a first implementation of this feature - I did make a number of refactors so it might be good if @drew08t / @ryantxu look them over to ensure I didn't miss something / test things locally to verify before merging in 👍 

Streaming even works :)

https://user-images.githubusercontent.com/22381771/191390973-4a8aa334-bfb9-4d13-8431-90d167303bd4.mov

**Comment by drew08t** [comment]
> This is looking good to me as a first implementation of this feature - I did make a number of refactors so it might be good if @drew08t / @ryantxu look them over to ensure I didn't miss something / test things locally to verify before merging in 👍
> 
> Streaming even works :)
> 
>  Fit.data.streaming.demo.mov

That streaming video is awesome!

**Comment by drew08t** [comment]
Refactoring and clean-up LGTM! Tested locally and works as expected!

**Pull request merged**
